c/misra/src/rules/RULE-13-2/UnsequencedSideEffects.ql

```diff
@@ -0,0 +1,135 @@
+  not variableEffect1.isPartial() and
+  variableEffect1.getAnAccess() = va1 and
+  (
+    exists(VariableEffect variableEffect2 |
+      not variableEffect2.isPartial() and
+      variableEffect2.getAnAccess() = va2 and
+      // If the effect is not local (happens in a different function) we use the call with the access as a proxy.
+      (
+        va1.getEnclosingStmt() = variableEffect1.getEnclosingStmt() and
+        va2.getEnclosingStmt() = variableEffect2.getEnclosingStmt() and
+        orderingConfig.isUnsequenced(variableEffect1, variableEffect2)
+        or
+        va1.getEnclosingStmt() = variableEffect1.getEnclosingStmt() and
+        not va2.getEnclosingStmt() = variableEffect2.getEnclosingStmt() and
+        exists(Call call |
+          call.getAnArgument() = va2 and call.getEnclosingStmt() = va1.getEnclosingStmt()
+        |
+          orderingConfig.isUnsequenced(variableEffect1, call)
+        )
+        or
+        not va1.getEnclosingStmt() = variableEffect1.getEnclosingStmt() and
+        va2.getEnclosingStmt() = variableEffect2.getEnclosingStmt() and
+        exists(Call call |
+          call.getAnArgument() = va1 and call.getEnclosingStmt() = va2.getEnclosingStmt()
+        |
+          orderingConfig.isUnsequenced(call, variableEffect2)
+        )
+      ) and
+      // Break the symmetry of the ordering relation by requiring that the first expression is located before the second.
+      // This builds upon the assumption that the expressions are part of the same full expression as specified in the ordering configuration.
+      exists(int offset1, int offset2 |
+        va1.getLocation().charLoc(_, offset1, _) and
+        va2.getLocation().charLoc(_, offset2, _) and
+        offset1 < offset2
+      ) and
+      placeHolder = variableEffect2 and
+      label = "side effect"
+    )
+    or
+    placeHolder = va2 and
+    label = "read" and
+    not exists(VariableEffect variableEffect2 | variableEffect1 != variableEffect2 |
+      variableEffect2.getAnAccess() = va2
+    ) and
+    (
+      va1.getEnclosingStmt() = variableEffect1.getEnclosingStmt() and
+      orderingConfig.isUnsequenced(variableEffect1, va2)
+      or
+      not va1.getEnclosingStmt() = variableEffect1.getEnclosingStmt() and
+      exists(Call call |
+        call.getAnArgument() = va1 and call.getEnclosingStmt() = va2.getEnclosingStmt()
+      |
+        orderingConfig.isUnsequenced(call, va2)
+      )
+    ) and
+    // The read is not used to compute the effect on the variable.
+    // E.g., exclude x = x + 1
+    not variableEffect1.getAChild+() = va2
+  ) and
+  // Both are evaluated
+  not exists(ConditionalExpr ce |
+    ce.getThen().getAChild*() = va1 and ce.getElse().getAChild*() = va2
+  )
+select fullExpr, "The expression contains unsequenced $@ to $@ and $@ to $@.", variableEffect1,
```

Comment:
Yes, in this example they are the same because the side effect is the variable access because the variable is `volatile`. In other cases, the side effects will be different and affect different variable access of the same variable.

Is "read" vs "side effect" unclear in the context of the whole alert message?
The goal is to find unsequenced side effects pairs or unsequenced side effect and variable read pairs.

Reply:
hmm yeah the rule talks about the read accesses, so I was thinking that was fine but the rule uses the language "modification" instead of side effect in the amplification,

so I was thinking it may be easier to imagine if the wording was more similar to the rule, with the 4 categories of how it could be violated

just a suggestion though! would be happy to also leave as is if you think that best describes what the rule finds!

Reply:
@knewbury01 @rvermeulen feel free to open an issue for this.